src/vmm/src/gdb/target.rs

```diff
@@ -0,0 +1,752 @@
+
+            r8: regs.regs[8],
+            r9: regs.regs[9],
+            r10: regs.regs[10],
+            r11: regs.regs[11],
+            r12: regs.regs[12],
+            r13: regs.regs[13],
+            r14: regs.regs[14],
+            r15: regs.regs[15],
+
+            rip: regs.rip,
+            rflags: regs.eflags as u64,
+        };
+
+        self.set_regs(new_regs, tid).map_err(|e| {
+            error!("Error setting registers {e:?}");
+            TargetError::NonFatal
+        })
+    }
+
+    /// Writes data to a guest virtual address for the Vcpu
+    fn read_addrs(
+        &mut self,
+        start_addr: <Self::Arch as Arch>::Usize,
+        data: &mut [u8],
+        tid: Tid,
+    ) -> TargetResult<usize, Self> {
+        let len = data.len();
+
+        let mut total_read = 0_usize;
+        let start_addr = usize::try_from(start_addr).map_err(|e| {
+            error!("Error {e:?} converting start addr to usize");
+            TargetError::NonFatal
+        })?;
+
+        while total_read < len {
+            let gaddr = start_addr + total_read;
+
+            let paddr = self.translate_gva(tid, gaddr as u64).map_err(|e| {
+                error!("Error {e:?} translating gva on read address: {start_addr:X}");
+                TargetError::NonFatal
+            })?;
+
+            let paddr = usize::try_from(paddr).map_err(|e| {
+                error!("Error {e:?} converting paddr to usize");
+                TargetError::NonFatal
+            })?;
+
+            let psize = arch::PAGE_SIZE;
+            // Compute the amount space left in the page after the paddr
+            let read_len = std::cmp::min(len - total_read, psize - (paddr & (psize - 1)));
+
+            let vmm = &self.vmm.lock().map_err(|_| {
+                error!("Error locking vmm in read addr");
+                TargetError::NonFatal
+            })?;
+            vmm.guest_memory()
+                .read(
+                    &mut data[total_read..total_read + read_len],
+                    GuestAddress(paddr as u64),
+                )
+                .map_err(|e| {
+                    error!("Error reading memory {e:?}");
+                    TargetError::NonFatal
+                })?;
+
+            total_read += read_len;
+        }
+
+        Ok(total_read)
+    }
+
+    /// Writes data at a guest virtual address for the Vcpu
+    fn write_addrs(
+        &mut self,
+        start_addr: <Self::Arch as Arch>::Usize,
+        data: &[u8],
+        tid: Tid,
+    ) -> TargetResult<(), Self> {
+        let len = data.len();
+
+        let mut total_written = 0_usize;
+        let start_addr = usize::try_from(start_addr).map_err(|e| {
+            error!("Error {e:?} unable to convert addr to usize");
+            TargetError::NonFatal
+        })?;
+
+        while total_written < len {
+            let gaddr = start_addr + total_written;
+            let paddr = match self.translate_gva(tid, gaddr as u64) {
+                Ok(0) => gaddr,
```

Comment:
Why does 0 here mean that we should treat the gva as a gpa? (ignoring the fact that due to the `NonZeroU64` stuff we'll never get 0 here anyway)